### PR TITLE
feat: Add KMS key and encrypt S3, SQS and SNS

### PR DESCRIPTION
**Pull request by MancunianSam**
I've made the KMS key argument mandatory for S3, SQS and SNS so that
they'll always be encrypted with KMS.

I've also removed some unused variables from the cloudwatch_alarms module.

### CODE REVIEW

File: kms/main.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
+resource "aws_kms_key" "encryption" {
+  description         = var.key_description
+  enable_key_rotation = true
+  policy              = var.key_policy
```
[the remark this answers is not in the document]
Comment (kurtismash) [reply kms/main.tf:4]: + is it worth this module creating a full key admin iam role itself?

File: kms/variables.tf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    persistent_resource_roles = []
+    ci_roles                  = []
+  }
+}
+
+variable "user_roles" {
+  default     = []
+  type        = list(string)
+  description = "A list of IAM roles who will have access to decrypt"
+}
```
Comment (kurtismash) [inline kms/variables.tf:38]: Is this now unused?

Comment (MancunianSam) [reply kms/variables.tf]: Yeah it is, I've taken it out.

File: kms/main.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (MancunianSam) [reply kms/main.tf]: This now creates a default policy. I've added a section in the README about what it does. The module also creates the key admin role as well with a trust policy set to deny all.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just a couple of minor editorial comments on the README

File: kms/main.tf [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  }
+  tags = merge(
+    var.tags,
+    tomap(
+      { "Name" = "${var.key_name}-admin" }
+    )
+  )
+}
+
+data "aws_iam_policy_document" "key_policy" {
```
Comment (kurtismash) [inline kms/main.tf:35]: Is it worth putting `Sid` properties in the statements to describe what they're doing for users of this module?

**Comment by MancunianSam** [comment]
This is up to date now. The KMS key is working for the S3 bucket -> SNS -> SQS -> Lambda for the Preservica config updates and for the Cloudwatch -> SNS -> SQS -> Lambda for the DLQ notifications. 
There may be some changes needed with other use cases but for now, it's working.

File: kms/main.tf [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+resource "aws_kms_alias" "encryption" {
+  name          = "alias/${var.key_name}"
+  target_key_id = aws_kms_key.encryption.key_id
+}
+
+module "kms_admin_role" {
+  source             = "../iam_role"
+  assume_role_policy = templatefile("${path.module}/templates/deny_all.json.tpl", {})
+  name               = "${var.key_name}-admin"
```
Comment (kurtismash) [inline kms/main.tf:23]: Should we prefix this with `KMS-` so it's clear when looking in the IAM console that this role is a key admin role?

File: README.md [hunk 5]
```diff
@@ -32,6 +32,18 @@ Creates an IAM policy from a string
...
+Creates A KMS key. 
+
+It also creates an admin role which has permissions to administer the key but not to decrypt or encrypt with it. 
+The trust policy of this role is set to deny access to everything. To assume this role, you will need to change the trust policy.
+
+The key policy allows the root user full KMS permissions. You can optionally provide lists of roles which will be added to the key policy.
+* User roles - these have permissions to decrypt and encrypt but not to make changes to the key.
+* CI roles - these have permissions to administer the key but not to delete it or use it to encrypt and decrypt.
+* Persistent resource roles - these have permissions to grant other AWS services access to the key.
+If these roles are not provided, the relevant policy statements won't be added.
```
Comment (kurtismash) [inline README.md:45]: Service names supported but not in README.

File: kms/main.tf [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        "kms:DescribeKey"
+      ]
+      effect    = "Allow"
+      resources = ["*"]
+      condition {
+        test     = "StringEquals"
+        values   = [data.aws_caller_identity.current.account_id]
+        variable = "aws:SourceAccount"
+      }
+    }
```
Comment (kurtismash) [inline kms/main.tf:165]: Missing `sid`

File: kms/main.tf (hunk 4 above)
Comment (MancunianSam) [reply kms/main.tf:23]: I'll do kms because most key names seem to have kebab case names

File: README.md (hunk 5 above)
Comment (MancunianSam) [reply README.md:45]: I've added a couple of lines to the README.

File: kms/main.tf (hunk 6 above)
Comment (MancunianSam) [reply kms/main.tf:165]: I've added one in.

File: kms/main.tf (hunk 4 above)
Comment (kurtismash) [reply kms/main.tf:23]: Why did this get reverted?

File: kms/main.tf [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
```
Comment (kurtismash) [inline kms/main.tf:166]: Will we need to do something special here for cross account access? E.g. subscribing an SQS to an encrypted SNS in another account?

File: sqs/main.tf
Comment (kurtismash) [inline sqs/main.tf:1]: We'll also need to encrypt the DLQs - if data in the main queue is encrypted it's unlikely to become less sensitive if it failed to process.

**Comment by kurtismash** [comment]
Looks almost ready for merging, just a few missing bits and a question.

File: kms/main.tf (hunk 7 above)
Comment (MancunianSam) [reply kms/main.tf:166]: I believe you can just change the source account which should allow services from other accounts.I've added a variable for source account which defaults to the calling account if it's not provided.

File: sqs/main.tf [hunk 8]
```diff
@@ -29,3 +52,19 @@ resource "aws_sqs_queue" "dlq" {
   name                      = "${var.queue_name}-dlq"
   message_retention_seconds = 1209600
 }
+
+module "dlq_metadata_and_files_cloudwatch_alarm" {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply sqs/main.tf]: Oops, I've changed the name.

File: sqs/main.tf
Comment (MancunianSam) [reply sqs/main.tf:1]: Fair point. I've changed it to use the same KMS key as the one used to encrypt the main queue. If you're using SSE-SQS then it uses that for the DLQ as well.

File: sqs/main.tf [hunk 9]
```diff
@@ -25,7 +27,55 @@ resource "aws_sqs_queue" "sqs_queue" {
...
+  message_retention_seconds = 1209600
+  sqs_managed_sse_enabled   = true
+}
+
+module "dlq_cloudwatch_alarm" {
+  count               = var.dlq_notification_topic == "" ? 0 : 1
+  source              = "../cloudwatch_alarms"
+  metric_name         = "ApproximateNumberOfMessagesVisible"
+  namespace           = "AWS/SQS"
+  name                = "${var.queue_name}-alarm"
```
Comment (kurtismash) [inline sqs/main.tf:72]: How about? `${var.queue_name}-dlq-alarm` We're likely to have multiple alarms on some queues.

**Review by kurtismash — COMMENTED** [review]
Just 1 last thing.

File: kms/outputs.tf
Comment (kurtismash) [inline kms/outputs.tf:1]: Is it worth outputting the alias as well?

Comment (MancunianSam) [reply kms/outputs.tf:1]: Can do yeah. I've added it to the outputs.

**Review by kurtismash — APPROVED** [review]

**Pull request merged**